README.md
Alertmanager Event Handler
==========================

Use Prometheus' Alertmanager to handle arbitrary events.  Inspired by the
event handlers in Nagios.

This works as a generic web hook in Alertmanager.  Alerts routed to this
event handler must have an annotation `handler` defined who's value is
the event handler that should be executed.  Optionally, the annotation
`handler` value may include arbitrary arguments passed to the event handler.

    <handler> [arg1, [arg2 ...]]

The configuration file for `am-event-handler` contains a hash of known
handlers which maps to a Go templated string.  This string builds the
executable and arguments that will run as the user running `am-event-handler`.
This allows the passing of arbitrary data with reasonable security restrictions
in place.  Example:

    ALERT PrometheusInstanceDown
      IF up{job="prometheus", environment="prod"} == 0
      FOR 5m
      LABELS {
        severity="event"
      }
      ANNOTATIONS {
        summary="{{$labels.instance}} Prometheus Instance Down",
        description="The Prometheus Instance {{$labels.instance}} is down",
        runbook="https://wiki/prometheus-instance-down",
        handler="restart-prom {{$labels.instance}}"
      }

This alert would be routed to event handler receiver in the Alertmanager
by the `severity` label.  The receiver configuration might look like the
below:

    # A list of notification receivers.
    receivers:
      - name: foobar
        webhook_configs:
          - url: http://host:port/

Where the `host:port` is where `am-event-handler` is running.

The `am-event-handler` requires a configuration file that will map the
handlers into an executable.  These are not run through a shell and shell
redirection will not work, but you can use this to execute a shell.  The
following is an example configuration:

    handlers:
      restart-prom: "remctl {{ index .Argv 0 }} prom-restart"

Note that the supplied arguments are stored in the `Argv` slice of strings.

Meta Handlers
-------------

Handlers are defined in the configuration file and can be any arbitrary,
non-whitespace containing string.  The handler selected matches the first
word of the `handler` annotation exactly.  There is no fancy logic there.

However, there are two meta handlers that can be defined in the configuration
that affect what will be executed.

* `default`: A handler of this name will be executed when no handler
  annotation is present, or the requested handler cannot be found.
* `all`: This handler is run for all alerts whether they have a handler
  annotation or not.  It will be run in addition to (and after) any
  matching handler the alert requests.

Templating
----------

In the configuration file the command to be executed can be templated as
in the example above.  In addition to Go's built in [text template][1]
functionality the following are made available.

Variables:

* `.Status`: `string` The status of the alert.  This should be either "firing"
  or "resolved".
* `.Labels`: `map[string]string`  A map of the labels applied to this alert.
* `.Annotations`: `map[string]string`  A map of the annotations applied to
  this alert.
* `.StartsAt`: `string` The RFC 3339 date the alert began as from the
  Alertmanager.
* `.EndsAt`: `string` The RFC 3339 date the alert ended as from the
  Alertmanager.  May be "0001-01-01T00:00:00Z" when the alert is in progress.
* `.GeneratorURL`: `string` The URL to the originating Prometheus server and
  graph.
* `.Argv`: `[]string` A slice of strings holding the arguments present in
  the `handler` annotation.  The handler annotation is whitespace delimited
  and this slice begins at the first argument and does not include the name
  of the specified handler.
* `.Json`: `string` A JSON representation of this alert.  This is a filtered
  representation of the alert and not the complete JSON as provided by the
  Alertmanager.
* `.Timestamp`: `string` A UTC timestamp in RFC 3339 format of when Alertmanager
  hit the am-event-handler with this alert.

Functions:

* `replace <string> <substring> <replacement>`:  This allows simple replacement
  based substitution of strings.  Perhaps dealing with escaping some pesky
  quotes.  As the templates are specified in YAML there is YAML escaping done
  on top of the Go string escaping before the string is parsed as a template.

Contributing
------------

Clone this repository and create a GitHub PR.

Authors and Copyright
---------------------

Copyright 2016 - 2017 42 Lines, Inc.  Original author: Jack Neely <jjneely@42lines.net>

[1]: https://golang.org/pkg/text/template/


debian/changelog
am-event-handler (0.0.7) unstable; urgency=medium

  * Add a 'replace' function to the templates

 -- Jack Neely <jjneely@42lines.net>  Mon, 10 Apr 2017 12:15:38 -0400

am-event-handler (0.0.6) unstable; urgency=medium

  * Handle arbitrarily sized alert payloads.

 -- Jack Neely <jjneely@42lines.net>  Tue, 06 Dec 2016 11:28:25 -0500

am-event-handler (0.0.5) unstable; urgency=medium

  * Log errors to the local log as well as report them to the remote client

 -- Jack Neely <jjneely@42lines.net>  Tue, 08 Nov 2016 10:14:23 -0500

am-event-handler (0.0.4) unstable; urgency=medium

  * Make the output JSON fields match the input fields

 -- Jack Neely <jjneely@42lines.net>  Mon, 24 Oct 2016 14:25:52 -0400

am-event-handler (0.0.3) unstable; urgency=medium

  * Add a timestamp to the alert metadata

 -- Jack Neely <jjneely@42lines.net>  Fri, 21 Oct 2016 15:25:59 -0400

am-event-handler (0.0.2) unstable; urgency=medium

  * Adding support for "default" and "all" special handlers

 -- Jack Neely <jjneely@42lines.net>  Thu, 20 Oct 2016 14:16:20 -0400

am-event-handler (0.0.1) unstable; urgency=medium

  * Initial Release.

 -- Jack Neely <jjneely@42lines.net>  Tue, 11 Oct 2016 16:57:28 -0400


argv_test.go
package main

import (
	"testing"
)

// argvParseTests is a map of test case => returned slice for the Tokenize
// function.
var argvParseTests = map[string][]string{
	"this is a test":           {"this", "is", "a", "test"},
	"this  is   a   test":      {"this", "is", "a", "test"},
	"this  is\na   test":       {"this", "is", "a", "test"},
	"this \"is a\"   test":     {"this", "is a", "test"},
	"this \"is\na\"   test":    {"this", "is\na", "test"},
	"this 'is a'   test":       {"this", "is a", "test"},
	"this 'is\na' \n\t  test":  {"this", "is\na", "test"},
	"this 'is \" a'   test":    {"this", "is \" a", "test"},
	"this \"is 'a\"   test":    {"this", "is 'a", "test"},
	"this \\\"is a\\\"   test": {"this", "\"is", "a\"", "test"},
	"this is\\\ta test":        {"this", "is\ta", "test"},
	"this \\is a test":         {"this", "\\is", "a", "test"},
	"this \\'is a test":        {"this", "'is", "a", "test"},
}

func equal(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}

	for i := 0; i < len(a); i++ {
		if a[i] != b[i] {
			return false
		}
	}

	return true
}

func TestTokenize(t *testing.T) {
	for k, v := range argvParseTests {
		tokens, err := Tokenize(k)
		if err != nil {
			t.Errorf("An unexpected error occured: %s", err.Error())
			continue
		}
		t.Logf("%s => %s", k, tokens)
		if !equal(v, tokens) {
			t.Errorf("%t != computed: %t", v, tokens)
		}
	}
}

func TestTokenizeError(t *testing.T) {
	// These strings do not successfully parse due to missing quotes
	var badStrings = []string{
		"this is a \"test",
		"this is a 'test",
	}

	for _, v := range badStrings {
		tokens, err := Tokenize(v)
		if err == nil {
			t.Errorf("This string should return an error: \"%s\" but it returned a successful parsed slice %s", v, tokens)
		} else {
			t.Logf("Error from bad string returned as expected: %s", err.Error())
		}
	}
}


main_test.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"strings"
	"testing"
	"time"
)

var testdata = map[string]int{
	"testdata/test1": 200, // No handler in annotation
	"testdata/test2": 200, // No handler in annotation
	"testdata/test3": 200, // Command not found on system (debug mode)
	"testdata/test4": 200, // Runs command on local system with templating
	"testdata/test9": 200, // Templating/quoting tests
}

var bind = "127.0.0.1:4242"

func init() {
	var err error

	// load test configuration into global config variable
	debug = true
	verbose = true
	timeout = time.Second * 15
	config, err = loadConfiguration("testdata/config.yaml")
	if err != nil {
		panic("Could not load test configuration: " + err.Error())
	}

	go run(bind)
	time.Sleep(1 * time.Second)
}

func TestREST(t *testing.T) {
	url := fmt.Sprintf("http://%s/", bind)
	buf := make([]byte, 4096)
	for k, v := range testdata {
		t.Logf("Testing %s", k)
		testcase := new(bytes.Buffer)
		json, err := os.Open(k)
		if err != nil {
			t.Fatal(err)
		}
		_, err = io.Copy(testcase, json)
		json.Close()
		if err != nil {
			t.Fatal(err)
		}

		resp, err := http.Get(url)
		if err != nil {
			t.Fatal(err)
		}
		if resp.StatusCode != 400 {
			// GETs are always Bad Requests
			t.Errorf("GET test returned status code %d", resp.StatusCode)
		}
		resp.Body.Close()
		resp, err = http.Post(url, "application/foobar", testcase)
		if err != nil {
			t.Fatal(err)
		}
		n, err := resp.Body.Read(buf)
		resp.Body.Close()
		if resp.StatusCode != v {
			t.Errorf("Bad Status from test: %d  Body: %s", resp.StatusCode,
				string(buf[:n]))
		}
	}
}

func postHelper(filename string) (*http.Response, error) {
	url := fmt.Sprintf("http://%s/", bind)
	testcase := new(bytes.Buffer)

	json, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	_, err = io.Copy(testcase, json)
	json.Close()
	if err != nil {
		return nil, err
	}

	return http.Post(url, "application/foobar", testcase)
}

func executeTest(t *testing.T, testcase, flagFile string) {
	// Holodeck safeties are off
	debug = false

	// Remove our test marker, ignoring errors
	_ = os.Remove(flagFile)

	resp, err := postHelper(testcase)
	if err != nil {
		t.Fatal(err)
	}

	buf := make([]byte, 4096)
	n, err := resp.Body.Read(buf)
	resp.Body.Close()
	if resp.StatusCode != 200 {
		t.Errorf("Bad Status from test: %d  Body: %s", resp.StatusCode,
			string(buf[:n]))
	}

	// Does our test file exist?
	_, err = os.Stat(flagFile)
	if os.IsNotExist(err) {
		t.Errorf("%s should exist after event handled, but does not", flagFile)
	} else if err != nil {
		t.Fatal(err)
	}

	_ = os.Remove(flagFile)
}

func TestExecution(t *testing.T) {
	executeTest(t, "testdata/test5", "testdata/unittest")
}

func TestDefaultHandler(t *testing.T) {
	config.Handlers["default"] = struct {
		Command string
		Status  string
	}{
		Command: "/bin/bash -c \"touch testdata/testDefault\"",
		Status:  "*",
	}
	executeTest(t, "testdata/test1", "testdata/testDefault")
	delete(config.Handlers, "default")
}

func TestAllHandler(t *testing.T) {
	config.Handlers["all"] = struct {
		Command string
		Status  string
	}{
		Command: "/bin/bash -c \"touch testdata/testAll\"",
		Status:  "*",
	}
	executeTest(t, "testdata/test1", "testdata/testAll")
	delete(config.Handlers, "all")
}

func TestTimeout(t *testing.T) {
	if testing.Short() {
		t.Skip("Skipping test in short mode.")
	}

	// Holodeck safeties are off
	debug = false

	resp, err := postHelper("testdata/test6")
	if err != nil {
		t.Fatal(err)
	}
	buf := make([]byte, 4096)
	n, err := resp.Body.Read(buf)
	resp.Body.Close()
	if resp.StatusCode != 400 {
		t.Errorf("Bad Status from test: %d  Body: %s", resp.StatusCode,
			string(buf[:n]))
	}

	t.Logf("Response body: %s", string(buf[:n]))
	if !strings.Contains(string(buf[:n]), "timed out") {
		t.Errorf("Timeout test did not return a timeout error")
	}
}

func TestJson(t *testing.T) {
	// Holodeck safeties are off
	debug = false

	resp, err := postHelper("testdata/test7")
	if err != nil {
		t.Fatal(err)
	}
	buf := make([]byte, 4096)
	n, err := resp.Body.Read(buf)
	resp.Body.Close()
	if resp.StatusCode != 200 {
		t.Errorf("Bad Status from test: %d  Body: %s", resp.StatusCode,
			string(buf[:n]))
	}

	t.Logf("Response body: %s", string(buf[:n]))
	alert := new(Alert)
	err = json.Unmarshal(buf[:n], alert)
	if err != nil {
		t.Fatalf("JSON unmarshalling failed")
	}

	// XXX: Does alert match what we expect?
}

func TestLargeSize(t *testing.T) {
	var body []byte

	if testing.Short() {
		t.Skip("Skipping test in short mode.")
	}

	// Holodeck safeties are on
	debug = true

	resp, err := postHelper("testdata/test8")
	if err != nil {
		t.Fatal(err)
	}
	buf := make([]byte, 4096)
	n := 0
	for err == nil {
		n, err = resp.Body.Read(buf)
		t.Logf("Read %d bytes.  IO error: %v", n, err)
		if n > 0 {
			body = append(body, buf[:n]...)
		}
	}
	if err != nil && err != io.EOF {
		t.Fatal(err)
	}
	resp.Body.Close()

	if resp.StatusCode != 200 {
		t.Errorf("Bad Status from test: %d", resp.StatusCode)
	}
}


argv.go
package main

import (
	"fmt"
	"unicode"
)

// State 0: No quote
// State 1: backslash
// State 2: In Quote

// Tokenize splits string s around each instance of one or more white space
// characters.  Unlike strings.Fields() it supports the use of shell-like
// quoting or tokenization to allow the returned substrings to contain white
// space.
func Tokenize(s string) ([]string, error) {
	var (
		result []string
		token  []rune
		state  int
		quote  rune
	)

	chunk := func() {
		if len(token) > 0 {
			result = append(result, string(token))
			token = nil // zero length slice
		}
	}

	for _, c := range s {
		switch state {
		case 0:
			switch {
			case unicode.IsSpace(c):
				chunk()
			case c == '\\':
				state = 1
			case c == '\'' || c == '"':
				state = 2
				quote = c
			default:
				token = append(token, c)
			}
		case 1:
			switch {
			case unicode.IsSpace(c):
				fallthrough
			case c == '\'':
				fallthrough
			case c == '"':
				token = append(token, c)
			default:
				token = append(token, '\\')
				token = append(token, c)
			}
			if quote == '\x00' {
				state = 0
			} else {
				state = 2
			}
		case 2:
			switch c {
			case '\\':
				state = 1
			case quote:
				chunk()
				quote = '\x00'
				state = 0
			default:
				token = append(token, c)
			}
		}
	}

	// End of String
	switch state {
	case 1:
		token = append(token, '\\')
	case 2:
		return nil, fmt.Errorf("Missing closing quote")
	}

	chunk()
	return result, nil
}


main.go
package main

import (
	"bytes"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"os/exec"
	"strings"
	"text/template"
	"time"

	"gopkg.in/yaml.v2"
)

const (
	// 4KiB buffer for the JSON body of the message
	JsonBody = 4096
)

// Errors
const (
	EMISSING = iota
)

var (
	// debug can be set to true to disable running of external commands
	debug bool

	// verbose can be set to true to enable more verbose logging
	verbose bool

	// timeout is the time we wait for each command run to complete before
	// canceling it.
	timeout time.Duration

	// config is a pointer to the global configuration object
	config *Configuration
)

// Alert represents an individual alert from Prometheus and included in the
// JSON blob POST'd via the Alertmanager.
type Alert struct {
	Status       string            `json:"status"`
	Labels       map[string]string `json:"labels"`
	Annotations  map[string]string `json:"annotations"`
	StartsAt     string            `json:"startsAt"`
	EndsAt       string            `json:"endsAt"`
	GeneratorURL string            `json:"generatorURL"`

	// Timestamp is a string representing the time Alertmanager hit this
	// API.  Useful for logging.
	Timestamp string `json:"timestamp"`

	// Argv is not in the alert JSON and is available so the handler arguments
	// can be exposed to the template.
	Argv []string `json:"-"`

	// Json is not from the alert JSON but holds a JSON formatted string
	// of this alert.  It is not the same JSON as originally passed in.
	Json string `json:"-"`
}

// AlertManagerEvent represents the JSON struct that is POST'd to a web_hook
// receiver from Prometheus' Alertmanager.  There are other fields in the
// JSON blob that are not included here.
type AlertManagerEvent struct {
	Version     string
	Status      string
	Receiver    string
	ExternalURL string
	Alerts      []Alert
}

// Configuration is the Golang type that represents the YAML structure of
// the configuration file.
type Configuration struct {
	// Handlers is a hash of handler name to the definition of what will
	// be executed.
	Handlers map[string]struct {

		// Command is the go template string of the command to execute
		Command string

		// Status is the status of the alert, either "firing" or "resolved",
		// that will trigger the handler execution.  A "*" character selects
		// any alert status.
		Status string
	}
}

// Error handling
type EventError struct {
	code   int
	object string
}

func (e EventError) Error() string {
	switch e.code {
	case EMISSING:
		return fmt.Sprintf("Handler %s is not defined or missing from configration",
			e.object)
	}

	return "Undefined event error"
}

// replace is a helper function for templating to do simple substitution.
func replace(a, b, c string) string {
	return strings.Replace(a, b, c, -1)
}

// loadConfiguration reads YAML data from the specified file name and populates
// a Configuration object.
func loadConfiguration(file string) (*Configuration, error) {
	fd, err := os.Open(file)
	if err != nil {
		return nil, err
	}
	defer fd.Close()

	body := make([]byte, JsonBody)
	size := 0
	size, err = fd.Read(body)
	if err != nil && err != io.EOF {
		return nil, err
	}

	cfg := new(Configuration)
	err = yaml.Unmarshal(body[:size], cfg)
	if err != nil {
		cfg = nil
	}
	return cfg, err
}

// formatHandler is a helper function to handle rendering the handler string
// templates.
func formatHandler(handler []string, command string, a Alert) (string, []string, error) {
	funcs := template.FuncMap{"replace": replace}
	// We ignore handler[0] as its the handle looked up to find command
	a.Argv = handler[1:]

	tmpl, err := template.New("command").Funcs(funcs).Parse(command)
	if err != nil {
		log.Printf("Error: Template parsing failed for \"%s\" with error: %s",
			command, err)
		return "", nil, err
	}
	buf := new(bytes.Buffer)
	err = tmpl.Execute(buf, a)
	if err != nil {
		log.Printf("Error: Template execution failed for \"%s\" with error: %s",
			command, err)
		return "", nil, err
	}

	// Tokenize here to preserve quoted arguments
	fields, err := Tokenize(buf.String())
	if err != nil {
		return "", nil, err
	}
	return fields[0], fields[1:], nil
}

// executeHandler executes a handler give an executable and a slice of
// arguments.  STDOUT and STDERR are merged together and returnd in the
// bytes.Buffer.
func executeHandler(exe string, args []string) (*bytes.Buffer, error) {
	done := make(chan error, 1)
	var err error
	if debug {
		log.Printf("DEBUG: Not executing command \"%s\" with args \"%#v\"", exe, args)
		return nil, nil
	}

	out := new(bytes.Buffer)
	cmd := exec.Command(exe, args...)
	cmd.Stderr = out
	cmd.Stdout = out
	start := time.Now().Unix()
	if err = cmd.Start(); err != nil {
		return nil, err
	}

	// This must be a channel to work with select() to implement a timeout
	go func() {
		done <- cmd.Wait()
	}()

	select {
	case err = <-done:
	case <-time.After(timeout):
		_ = cmd.Process.Kill() // Ignore error here
		err = fmt.Errorf("Command execution timed out and was killed.")
		out = nil
	}

	end := time.Now().Unix()
	if err != nil {
		log.Printf("Command \"%s\" Args \"%#v\" failed in %d seconds: %s",
			exe, args, end-start, err.Error())
	} else {
		log.Printf("Command \"%s\" Args \"%#v\" ran successfully in %d seconds",
			exe, args, end-start)
	}

	return out, err
}

// handleEvent does the initial work to handle events from the HTTP body.
func handleEvent(e *AlertManagerEvent) (*bytes.Buffer, error) {
	errors := 0
	retText := new(bytes.Buffer)
	for _, alert := range e.Alerts {
		log.Printf("Processing Alert: %s", alert.Labels["alertname"])
		var handler []string
		alert.Timestamp = time.Now().UTC().Format(time.RFC3339)

		buf, err := json.Marshal(alert)
		if err != nil {
			msg := fmt.Sprintf("Error marshalling JSON: %s", err.Error())
			log.Print(msg)
			retText.WriteString(msg + "\n")
			errors++
			continue
		}
		alert.Json = string(buf)
		if _, ok := alert.Annotations["handler"]; !ok {
			// We didn't find the "handler" annotation
			log.Printf("%s does not have handler annotation trying default",
				alert.Labels["alertname"])
			handler = []string{"default"}
		} else {
			handler = strings.Fields(alert.Annotations["handler"])
		}

		// Run our handler or the default if no handler is present.  Following
		// that run the "all" handler if present.
		for _, h := range [][]string{handler, []string{"all"}} {
			output, err := parseHandler(h, alert)
			if err != nil {
				if e, ok := err.(EventError); ok && e.code == EMISSING {
					if h[0] == "default" || h[0] == "all" {
						// Ignore missing handler errors for our special handlers
						// This means that a missing handler annotation is not
						// considered an error.
						continue
					}
				}
				log.Printf(err.Error())
				retText.WriteString(err.Error() + "\n")
				errors++
			}
			if output != nil && output.Len() > 0 {
				retText.Write(output.Bytes())
			}
		}
	}

	if errors > 0 {
		return retText, fmt.Errorf("Error(s) executing event(s)")
	}

	return retText, nil
}

// parseHandler parses and error checks the handler string before execution.
func parseHandler(handler []string, alert Alert) (*bytes.Buffer, error) {
	if len(handler) == 0 {
		return nil, fmt.Errorf("Empty handler annotation found in alert.")
	}
	command, ok := config.Handlers[handler[0]]
	if !ok {
		return nil, EventError{EMISSING, handler[0]}
	}
	if command.Status == "" {
		// Set default value for non-specified status
		command.Status = "firing"
	}
	if command.Status != "*" && command.Status != alert.Status {
		log.Printf("Ignoring alert.  Status (%s) which does not match filter (%s)",
			alert.Status, command.Status)
		return nil, nil
	}
	script, args, err := formatHandler(handler, command.Command, alert)
	if err != nil {
		return nil, fmt.Errorf("Could not parse handler arguments: %s", err.Error())
	}
	if script == "" {
		// Sanity
		return nil, fmt.Errorf("Script is empty, not running.")
	}

	return executeHandler(script, args)
}

// unmarshalBody is a helper function to load JSON from an HTTP body into
// an AlertManagerEvent structure.
func unmarshalBody(encoded []byte) (*AlertManagerEvent, error) {
	data := new(AlertManagerEvent)
	err := json.Unmarshal(encoded, &data)
	if err != nil {
		return nil, err
	}

	return data, nil
}

// amWebHook decodes the HTTP request, finds Alertmanager JSON structure
// and dispatches the alerts.
func amWebHook(writer http.ResponseWriter, r *http.Request) {
	var body []byte
	var err error
	var n int

	// Log the request
	w := NewStatusResponseWriter(writer)
	defer logRequest(w, r)

	// Filter requests for POST
	if r.Method != "POST" {
		http.Error(w, "Bad request method.", http.StatusBadRequest)
		return
	}

	buf := make([]byte, JsonBody)
	for err == nil {
		n, err = r.Body.Read(buf)
		if err != nil && err != io.EOF {
			log.Printf("Error reading from client: %s", err.Error())
			http.Error(w, err.Error(), http.StatusInternalServerError)
			return
		}
		if n > 0 {
			body = append(body, buf[:n]...)
		}
	}

	if verbose {
		log.Printf("Request Body: \"%s\"", string(body))
	}

	event, err := unmarshalBody(body)
	if err != nil {
		log.Printf("Error parsing request JSON: %s", err.Error())
		http.Error(w, "Error parsing JSON: "+err.Error(), http.StatusBadRequest)
		return
	}

	output, err := handleEvent(event)
	if err != nil {
		w.WriteHeader(http.StatusBadRequest)
	} else {
		w.WriteHeader(http.StatusOK)
	}
	if output.Len() > 0 {
		blob := output.Bytes()
		w.Write(blob)
		if verbose {
			log.Printf("Response body: %s", string(blob))
		}
	}
}

// run starts the HTTP server
func run(bindAddress string) {
	http.HandleFunc("/", amWebHook)

	log.Printf("Starting server on %s", bindAddress)
	err := http.ListenAndServe(bindAddress, nil)
	if err != nil {
		log.Fatal(err)
	}
}

func main() {
	var bindAddress string
	var configFile string
	var err error

	flag.StringVar(&bindAddress, "bind", "0.0.0.0:4242",
		"IP:PORT to listen for HTTP requests.")
	flag.StringVar(&bindAddress, "b", "0.0.0.0:4242",
		"IP:PORT to listen for HTTP requests.")
	flag.StringVar(&configFile, "config", "./config.yaml",
		"Configuration file.")
	flag.StringVar(&configFile, "c", "./config.yaml",
		"Configuration file..")
	flag.BoolVar(&debug, "debug", false, "Activate debug mode.")
	flag.BoolVar(&debug, "d", false, "Activate debug mode.")
	flag.BoolVar(&verbose, "verbose", false, "Verbose logging.")
	flag.BoolVar(&verbose, "v", false, "Verbose logging.")
	flag.DurationVar(&timeout, "timeout", time.Second*30, "Command/Handler timeout.")
	flag.DurationVar(&timeout, "t", time.Second*30, "Command/Handler timeout.")

	flag.Parse()
	config, err = loadConfiguration(configFile)
	if err != nil {
		log.Fatalf("Configuration error, aborting: %s", err)
	}
	for k, v := range config.Handlers {
		log.Printf("Found handler %s => %s", k, v)
	}

	run(bindAddress)
}
